### PR TITLE
WIP STLC example

### PR DESCRIPTION
**Pull request by thumphries**
This could use a fair bit of cleaning up, but is mostly there. The generators are working reasonably well, though they may not be idiomatic. `genWellTypedApp` is producing reasonably interesting chains of function application.

As you can see, I haven't bothered implementing `subst`, should probably do that.

@jystic @charleso

### CODE REVIEW

File: hedgehog-example/test/Test/Example/STLC.hs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+  | EInt Int
+  | EString Text
+  | EVar Text
+  | ELam Text Type Expr
+  | EApp Expr Expr
+  deriving (Eq, Ord, Show)
+
+-- -----------------------------------------------------------------------------
+
+-- | Evaluate to weak head normal form.
```
Comment (thumphries) [inline hedgehog-example/test/Test/Example/STLC.hs]: would probably get more interesting failures if this went all the way to normal form, was just being lazy.

**Comment by thumphries** [comment]
Not sure how to minimise this so it's a good teaching example, just threw in most of the techniques I've found useful.

File: hedgehog-example/test/Test/Example/STLC.hs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+  eg <- genWellTypedExpr' tg
+  let tf = TArrow tg want
+  ef <- genWellTypedExpr' tf
+  pure (EApp ef eg)
+
+-- | This tries to look up a known expression of the desired type from the env.
+-- It does not always succeed, throwing `empty` when unavailable.
+genWellTypedPath :: Type -> Gen (Reader (Map Type [Expr])) Expr
+genWellTypedPath want = do
+  paths <- lift ask
```
Comment (jystic) [inline hedgehog-example/test/Test/Example/STLC.hs]: You should be able to replace this with just `ask` now we have `MonadReader`

**Comment by jystic** [comment]
I'm going to merge this so I can keep it up to date with my churning

**Comment by jystic** [comment]
I think there's no harm having it in the examples area, happy for it to be cleaned up whenever you have time.

**Pull request merged**